feat(opentelemetry): Ensure DSC & attributes are correctly set

mydea opened the pull request:
This PR fixes a few things about the OTEL handling of our SDK.

1. Ensure that attributes are actually correctly put on the trace data - we used to remove e.g. `sentry.sample_rate` which we don't do in "main" SDK. Question: Should we remove the `otel.attributes` context data then? As this is kind of redundant at this point with the `data` on the trace?
2. Ensure we put the `transaction` tag correctly on the transaction, to align with "main" SDK.
3. Ensure we store the incoming DSC as traceState, so we can pick this up. This can include e.g. transaction name and stuff like this to persist.

Lms24 reviewed (approved):
LGTM!

> Should we remove the otel.attributes context data then?

What field is this exactly on the event? I don't think we should remove stuff from `event.contexts.trace.data` (which IIUC span attributes end up in).

mydea commented:
> LGTM!
> 
> > Should we remove the otel.attributes context data then?
> 
> What field is this exactly on the event? I don't think we should remove stuff from `event.contexts.trace.data` (which IIUC span attributes end up in) because IIRC either ingest or the product later need these fields.

What I mean is that we send the "root span" attributes in `contexts.trace.data`, but we also send all otel span attributes in `contexts.otel.attributes` - which now is basically the same list of attributes, twice 🤔

lforst commented on packages/opentelemetry/src/utils/dynamicSamplingContext.ts:
```diff
@@ -0,0 +1,61 @@
+    return {};
+  }
+
+  const dsc = getDynamicSamplingContextFromClient(span.spanContext().traceId, client);
+
+  const traceState = span.spanContext().traceState;
+  const traceStateDsc = traceState?.get(SENTRY_TRACE_STATE_DSC);
+
+  if (traceStateDsc) {
+    const dsc = baggageHeaderToDynamicSamplingContext(traceStateDsc);
+    if (dsc) {
+      return dsc;
+    }
```
`dsc` is defined twice in this scope. Can we somehow pull them apart so that it is more clear what the order of priority is in this function? I personally would first do the base-case and check whether there is a tracestate/dsc on the span and return that one. If it's not there we should do the rest of the logic.

mydea replied:
good point, will rewrite this a bit!

lforst reviewed (commented):
I have a feeling that we are missing the inverse logic in `inject`. More concretely we need to take the dynamic sampling context of the current span when inject is called, e.g. by calling `getDynamicSamplingContextFromSpan([current span])` on the current span. I think I might have missed this previously.

mydea commented on packages/opentelemetry/src/setupEventContextTrace.ts:
```diff
@@ -1,28 +1,37 @@
     // If event has already set `trace` context, use that one.
     event.contexts = {
-      trace: {
+      trace: dropUndefinedKeys({
```
I figured it's nicer to not send an undefined `parent_span_id` (this was one debugging step when I was fixing tests etc), but we can also leave it!

mydea commented on packages/opentelemetry/src/spanExporter.ts:
```diff
@@ -159,13 +171,13 @@ function createTransactionForOtelSpan(span: ReadableSpan): Transaction {
     startTimestamp: convertOtelTimeToSeconds(span.startTime),
     metadata: {
-      sampleRate: span.attributes[SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE] as number | undefined,
+      ...dropUndefinedKeys({
```
here I think it is important, because `dynamicSamplingContext` may be `undefined` in which case we don't want to set it, because later in the pipeline if you have e.g. this:

```
metadata: {
  dynamicSamplingContext: xx,
  ...event.metadata
}
```
this would overwrite `xx` even if it is `undefined`, I think?

mydea commented:
OK, I also updated the propagator to look at the span trace state first, and only then fall back to propagation context! I also added tests to cover this, and also to ensure trace state is correctly inherited & applied for spans.
